### PR TITLE
feat: add Constant Thrust Thruster Dynamics 

### PR DESCRIPTION
**Pull request by Derollez**
- Add `Thruster` and `ConstantThrust` Dynamics. 
- Add validation tests against Orekit (with drag and without drag) --> Check validation error thresholds on the different cases in `Propagator.test.cpp`. 
- Python bindings. 
- Python tutorial notebooks.

### CODE REVIEW

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /// @brief              Print thruster
+    ///
+    /// @param              [in] anOutputStream An output stream
+    /// @param              [in] (optional) displayDecorators If true, display decorators
+
+    virtual void print(std::ostream& anOutputStream, bool displayDecorator = true) const;
+
+   private:
+    const String name_;
+    SatelliteSystem satelliteSystem_;  // TBI: Might want to use PropulsionSystem instead and populate it under the hood
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp:91]: I think it's worth keeping the satellite system as we might want to separate dry mass and wet mass.
We can track the dry mass in the satellite system. And the wet mass in the State.

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+{
+   public:
+    /// @brief              Constructor
+    ///
+    /// @code
+    ///                     const String aName = { ... };
+    ///                     const SatelliteSystem aSatelliteSystem = { ... };
+    ///                     const Direction aThrustDirection = { ... };
+    ///                     ConstantThrustThruster constantThrustThrusterDynamics = { aName, aSatelliteSystem,
+    ///                     aThrustDirection};
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:59]:
```suggestion
    ///                     const SatelliteSystem aSatelliteSystem = { ... };
    ///                     const Direction aThrustDirection = { ... };
    ///                     const String aName = { ... };
    ///                     ConstantThrustThruster constantThrustThrusterDynamics = {aSatelliteSystem,
    ///                     aThrustDirection, aName};
```

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+using ostk::astro::flight::system::SatelliteSystem;
+using ostk::astro::flight::system::PropulsionSystem;
+using ostk::astro::flight::system::Dynamics;
+using ostk::astro::trajectory::State;
+using ostk::astro::trajectory::LocalOrbitalFrameDirection;
+
+/// @brief                      Define the acceleration experienced by a point mass due to a constant thrust thruster
+/// dynamics
+
+class ConstantThrustThruster : public Thruster
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:49]: All the docstrings below have to be aligned with the above (1 tab)

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    ///
+    /// @code
+    ///                     const String aName = { ... };
+    ///                     const SatelliteSystem aSatelliteSystem = { ... };
+    ///                     const Direction aThrustDirection = { ... };
+    ///                     ConstantThrustThruster constantThrustThrusterDynamics = { aName, aSatelliteSystem,
+    ///                     aThrustDirection};
+    /// @endcode
+    ///
+    /// @param              [in] aDirection A direction
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:62]:
```suggestion
    /// @param              [in] aSatelliteSystem A satellite system
    /// @param              [in] aDirection A direction
```

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp (hunk 3 above)
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:49]:
```suggestion
class ConstantThrust : public Thruster
```
Nit: Maybe rename to ConstantThrust?

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        std::ostream& anOutputStream, const ConstantThrustThruster& aConstantThrustThruster
+    );
+
+    /// @brief              Check if constant thrust thruster dynamics is defined
+    ///
+    /// @return             True if constant thrust thruster dynamics is defined
+
+    virtual bool isDefined() const override;
+
+    /// @brief              Get constant thrust thruster dynamics' thrust
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:110]:
```suggestion
    /// @brief              Get thrust 
```
Nit: The brief is a bit cyclical.

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+namespace flight
+{
+namespace system
+{
+namespace dynamics
+{
+
+Thruster::Thruster(const SatelliteSystem& aSatelliteSystem, const String& aName)
+    : name_(aName),
+      satelliteSystem_(aSatelliteSystem)
```
Comment (vishwa2710) [inline src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.cpp:21]:
```suggestion
    : satelliteSystem_(aSatelliteSystem),
      name_(aName)
```

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp (hunk 1 above)
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp:91]:
```suggestion
    SatelliteSystem satelliteSystem_;  // TBI: Might want to use PropulsionSystem instead and populate it under the 
    hood
    String name_;
```

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+VectorXd ConstantThrustThruster::computeContribution(
+    const Instant& anInstant, const VectorXd& x, [[maybe_unused]] const Shared<const Frame>& aFrameSPtr
+) const
+{
+    const Vector3d positionCoordinates = {x[0], x[1], x[2]};
+    const Vector3d velocityCoordinates = {x[3], x[4], x[5]};
+
+    const Position position_GCRF = {positionCoordinates, Position::Unit::Meter, Frame::GCRF()};
+    const Velocity velocity_GCRF = {velocityCoordinates, Velocity::Unit::MeterPerSecond, Frame::GCRF()};
```
Comment (vishwa2710) [inline src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:94]:
```suggestion
    const Position position_GCRF = {positionCoordinates, Position::Unit::Meter, aFrameSPtr};
    const Velocity velocity_GCRF = {velocityCoordinates, Velocity::Unit::MeterPerSecond, aFrameSPtr};
```

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    return {CartesianPosition::Default(), CartesianVelocity::Default(), CoordinatesSubset::Mass()};
+}
+
+Array<Shared<const CoordinatesSubset>> ConstantThrustThruster::getWriteCoordinatesSubsets() const
+{
+    return {CartesianVelocity::Default(), CoordinatesSubset::Mass()};
+}
+
+VectorXd ConstantThrustThruster::computeContribution(
+    const Instant& anInstant, const VectorXd& x, [[maybe_unused]] const Shared<const Frame>& aFrameSPtr
```
Comment (vishwa2710) [inline src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:87]:
```suggestion
    const Instant& anInstant, const VectorXd& x, const Shared<const Frame>& aFrameSPtr
```

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    const Position position_GCRF = {positionCoordinates, Position::Unit::Meter, Frame::GCRF()};
+    const Velocity velocity_GCRF = {velocityCoordinates, Velocity::Unit::MeterPerSecond, Frame::GCRF()};
+
+    // Get Rotation Matrix from Direction Local Orbital Frame (LOF) to GCRF (ECI)
+    Shared<const Frame> frameSPtr = this->localOrbitalFrameDirection_.accessLocalOrbitalFrameFactory()->generateFrame(
+        anInstant, positionCoordinates, velocityCoordinates
+    );
+    Quaternion q_GCRF_LOF = frameSPtr->getTransformTo(Frame::GCRF(), anInstant).getOrientation();
+
+    // TBImproved
```
Comment (vishwa2710) [inline src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:102]:
```suggestion
    // TBM: Improve later
```

**Review by vishwa2710 — COMMENTED** [review]
Preliminary review. Let me know when this is ready, I thought it was but I feel like you're still adding content.

File: include/OpenSpaceToolkit/Astrodynamics/Trajectory/LocalOrbitalFrameDirection.hpp [hunk 10]
```diff
@@ -28,6 +28,8 @@ using ostk::astro::trajectory::LocalOrbitalFrameFactory;
 ///
 /// @note                       A unit vector, expressed with a Local Orbital Frame Factory.
 
+// Add docstrings
+
```
Comment (Derollez) [inline include/OpenSpaceToolkit/Astrodynamics/Trajectory/LocalOrbitalFrameDirection.hpp:32]:
```suggestion
```

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp (hunk 1 above)
Comment (Derollez) [reply include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster.hpp:91]: Yeah agreed, implemented this way

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp (hunk 3 above)
Comment (Derollez) [reply include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:49]: @phc1990 @kyle-cochran Thoughts?

File: test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        std::make_shared<EarthGravitationalModel>(EarthGravitationalModel::Type::Undefined),
+        std::make_shared<EarthMagneticModel>(EarthMagneticModel::Type::Undefined),
+        std::make_shared<EarthAtmosphericModel>(EarthAtmosphericModel::Type::Exponential),
+    };
+
+    const Shared<const Frame> gcrfSPtr_ = Frame::GCRF();
+
+    LocalOrbitalFrameDirection localOrbitalFrameDirection_ = LocalOrbitalFrameDirection::Undefined();
+
+    Mass satelliteDryMass_ = Mass::Undefined();
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp:145]:
```suggestion
    const Mass satelliteDryMass_ = Mass::Kilograms(95.0);
    const Mass satelliteWetMass_ = Mass::Kilograms(5.0);
```

File: test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    void SetUp() override
+    {
+        const Composite satelliteGeometry(Cuboid(
+            {0.0, 0.0, 0.0},
+            {Vector3d {1.0, 0.0, 0.0}, Vector3d {0.0, 1.0, 0.0}, Vector3d {0.0, 0.0, 1.0}},
+            {1.0, 2.0, 3.0}
+        ));
+
+        const Scalar thrust_ = Scalar(0.1, PropulsionSystem::thrustSIUnit);
+        const Scalar specificImpulse_ = Scalar(1500.0, PropulsionSystem::specificImpulseSIUnit);
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp:98]:
```suggestion
```

File: test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+        satelliteDryMass_ = Mass::Kilograms(100.0);
+
+        propulsionSystem_ = PropulsionSystem(
+            thrust_,          // Thrust
+            specificImpulse_  // Isp
+        );
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp:105]:
```suggestion
```

File: test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+        std::make_shared<EarthMagneticModel>(EarthMagneticModel::Type::Undefined),
+        std::make_shared<EarthAtmosphericModel>(EarthAtmosphericModel::Type::Exponential),
+    };
+
+    const Shared<const Frame> gcrfSPtr_ = Frame::GCRF();
+
+    LocalOrbitalFrameDirection localOrbitalFrameDirection_ = LocalOrbitalFrameDirection::Undefined();
+
+    Mass satelliteDryMass_ = Mass::Undefined();
+    PropulsionSystem propulsionSystem_ = PropulsionSystem::Undefined();
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.test.cpp:146]:
```suggestion
    PropulsionSystem propulsionSystem_ = {
        Scalar(0.1, PropulsionSystem::thrustSIUnit),
        Scalar(1500.0, PropulsionSystem::specificImpulseSIUnit),
    };
```

**Review by vishwa2710 — COMMENTED** [review]
Looks great man!

File: test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp [hunk 15]
```diff
@@ -1901,6 +1940,1056 @@ TEST_F(OpenSpaceToolkit_Astrodynamics_Trajectory_Orbit_Models_Propagator, PropAc
     }
 }
 
+// TBI: Specific class for parameterized tests on thruster dynamics, could inherit from base fixture for common inputs
+// error: 'testing::Test' is an ambiguous base
```
[the remark this answers is not in the document]
Comment (Derollez) [reply test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp:1944]: Yup, initially tried with this, there is something else with this situation

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/AtmosphericDrag.cpp [hunk 16]
```diff
@@ -126,11 +127,11 @@ VectorXd AtmosphericDrag::computeContribution(
 
     const Vector3d relativeVelocity = velocityCoordinates - earthAngularVelocity.cross(positionCoordinates);
 
-    const Real mass = satelliteSystem_.getMass().inKilograms();  // TBI: Add wet mass from state vector
+    const Real mass = x[6];
```
[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/AtmosphericDrag.cpp:130]: So the current implementation takes into account x[6] as the full satellite mass (i.e. wet mass)

File: include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp (hunk 3 above)
Comment (Derollez) [reply include/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.hpp:49]: Yeah okay good point. My view was that `ConstantThrust` alone (and same for dynamics children) does not necessarily convey the meaning of `Thruster` (or `Dynamics` for the others). But that can be made clear in the tutorials I think

**Review by phc1990 — COMMENTED** [review]
Left some comments, one of them of importance (marked as ⚠️ )

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/AtmosphericDrag.cpp [hunk 17]
```diff
@@ -126,11 +127,11 @@ VectorXd AtmosphericDrag::computeContribution(
...
     const Real dragCoefficient = satelliteSystem_.getDragCoefficient();
     const Real surfaceArea = satelliteSystem_.getCrossSectionalSurfaceArea();
 
-    // Add object's gravity to total gravitational acceleration
+    // Compute drag contribution to state derivative
     const Vector3d dragAccelerationSI =
         -(0.5 / mass) * dragCoefficient * surfaceArea * atmosphericDensity * relativeVelocity.norm() * relativeVelocity;
```
[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/AtmosphericDrag.cpp:136]: Yeah good point. I'll leave it for this PR but let's track and revisit later 💯

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+VectorXd ConstantThrustThruster::computeContribution(
+    const Instant& anInstant, const VectorXd& x, const Shared<const Frame>& aFrameSPtr
+) const
+{
+    const Vector3d positionCoordinates = {x[0], x[1], x[2]};
+    const Vector3d velocityCoordinates = {x[3], x[4], x[5]};
+
+    // Get Rotation Matrix from Direction Local Orbital Frame (LOF) to Requested Frame
+    Shared<const Frame> frameSPtr = this->localOrbitalFrameDirection_.accessLocalOrbitalFrameFactory()->generateFrame(
```
[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:101]: Done

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        anInstant, positionCoordinates, velocityCoordinates
```
[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:102]: Yeah agreed. Done. Leaving as is for a first release for now

File: bindings/python/test/flight/system/dynamics/test_atmospheric_drag.py [hunk 20]
```diff
@@ -60,12 +65,22 @@ def dynamics(earth: Earth, satellite_system: SatelliteSystem) -> AtmosphericDrag
...
+            CartesianVelocity.default(),
+            CoordinatesSubset.mass(),
+        ]
+    )
+
+
+@pytest.fixture
+def state(coordinates_broker: CoordinatesBroker) -> State:
+    instant: Instant = Instant.date_time(DateTime(2021, 3, 20, 12, 0, 0), Scale.UTC)
+    coordinates: list = [7000000.0, 0.0, 0.0, 0.0, 7546.05329, 0.0, 100.0]
```
[the remark this answers is not in the document]
Comment (Derollez) [reply bindings/python/test/flight/system/dynamics/test_atmospheric_drag.py:81]: yeah good point

File: src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    const SatelliteSystem& aSatelliteSystem, const LocalOrbitalFrameDirection& aThrustDirection, const String& aName
+)
+    : Thruster(aSatelliteSystem, aName),
+      localOrbitalFrameDirection_(aThrustDirection)
+{
+}
+
+ConstantThrustThruster::~ConstantThrustThruster() {}
+
+ConstantThrustThruster* ConstantThrustThruster::clone() const
```
[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:49]: Agreed

[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Flight/System/Dynamics/Thruster/ConstantThrustThruster.cpp:49]: Actually removed as not used and allowing to beef up the python bindings on top

File: test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp [hunk 22]
```diff
@@ -118,12 +132,23 @@ class OpenSpaceToolkit_Astrodynamics_Trajectory_Orbit_Models_Propagator : public
             {1.0, 2.0, 3.0}
         ));
 
+        const PropulsionSystem propulsionSystem = {
+            Scalar(0.1, PropulsionSystem::thrustSIUnit),
+            Scalar(1500.0, PropulsionSystem::specificImpulseSIUnit),
+        };
+
+        this->satelliteDryMass_ = Mass(100.0, Mass::Unit::Kilogram);
```
[the remark this answers is not in the document]
Comment (Derollez) [reply test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp:140]: Using it later in functions

File: test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp [hunk 23]
```diff
@@ -162,9 +187,19 @@ class OpenSpaceToolkit_Astrodynamics_Trajectory_Orbit_Models_Propagator : public
     const Shared<const Frame> gcrfSPtr_ = Frame::GCRF();
 
     Array<Shared<Dynamics>> defaultDynamics_ = Array<Shared<Dynamics>>::Empty();
+    Mass satelliteDryMass_ = Mass::Undefined();
```
[the remark this answers is not in the document]
Comment (Derollez) [reply test/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.test.cpp:190]: Okay, my bad, those 2 go together 💹

**Review by phc1990 — APPROVED** [review]
LGTM

**Pull request merged**